source/common/network/listener_impl.cc

```diff
@@ -43,85 +41,92 @@ bool ListenerImpl::rejectCxOverGlobalLimit() {
+    // for Unix domain sockets; apparently there isn't a mechanism in the kernel to get the
+    // `sockaddr_un` associated with the client socket when starting from the server socket.
+    // We work around this by using our own name for the socket in this case.
+    // Pass the 'v6only' parameter as true if the local_address is an IPv6 address. This has no
+    // effect if the socket is a v4 socket, but for v6 sockets this will create an IPv4 remote
+    // address if an IPv4 local_address was created from an IPv6 mapped IPv4 address.
+    const Address::InstanceConstSharedPtr& remote_address =
+        (remote_addr.ss_family == AF_UNIX)
+            ? io_handle->peerAddress()
+            : Address::addressFromSockAddr(remote_addr, remote_addr_len,
+                                           local_address->ip()->version() ==
+                                               Address::IpVersion::v6);
+
+    cb_.onAccept(
+        std::make_unique<AcceptedSocketImpl>(std::move(io_handle), local_address, remote_address));
   }
-
-  // Get the local address from the new socket if the listener is listening on IP ANY
-  // (e.g., 0.0.0.0 for IPv4) (local_address_ is nullptr in this case).
-  const Address::InstanceConstSharedPtr& local_address =
-      listener->local_address_ ? listener->local_address_ : io_handle->localAddress();
-
-  // The accept() call that filled in remote_addr doesn't fill in more than the sa_family field
-  // for Unix domain sockets; apparently there isn't a mechanism in the kernel to get the
-  // `sockaddr_un` associated with the client socket when starting from the server socket.
-  // We work around this by using our own name for the socket in this case.
-  // Pass the 'v6only' parameter as true if the local_address is an IPv6 address. This has no effect
-  // if the socket is a v4 socket, but for v6 sockets this will create an IPv4 remote address if an
-  // IPv4 local_address was created from an IPv6 mapped IPv4 address.
-  const Address::InstanceConstSharedPtr& remote_address =
-      (remote_addr->sa_family == AF_UNIX)
-          ? io_handle->peerAddress()
-          : Address::addressFromSockAddr(*reinterpret_cast<const sockaddr_storage*>(remote_addr),
-                                         remote_addr_len,
-                                         local_address->ip()->version() == Address::IpVersion::v6);
-  listener->cb_.onAccept(
-      std::make_unique<AcceptedSocketImpl>(std::move(io_handle), local_address, remote_address));
 }
 
 void ListenerImpl::setupServerSocket(Event::DispatcherImpl& dispatcher, Socket& socket) {
-  listener_.reset(
-      evconnlistener_new(&dispatcher.base(), listenCallback, this, 0, -1, socket.ioHandle().fd()));
-
-  if (!listener_) {
-    throw CreateListenerException(
-        fmt::format("cannot listen on socket: {}", socket.localAddress()->asString()));
-  }
+#ifdef WIN32
+  auto trigger = Event::FileTriggerType::Level;
+#else
+  auto trigger = Event::FileTriggerType::Edge;
```

Comment:
I think listeners should always operate on Level mode.

Operating in level mode allows us to safely accept only the first N connections available and return without having to set explicit resumption conditions.

Reply:
This is related to the comment above. We currently consume all the accepted sessions, before returning. If we plan to add a more sophisticated admissions mechanism, then we might want to switch.

Reply:
Use of level mode here always feels slightly safer.  I agree that use of edge trigger should mostly be ok since we drain to completion.  I worry about some transient error on accept resulting in us losing the trigger and then never recovering.  Plus, doing the same on all platforms is simpler.

Reply:
Fixed!